### ISSUE TITLE
Deprecate `glob` in favour of `globset` and `globwalk`?

### ISSUE THREAD

**Issue by KodrAus** [rust-lang-nursery/ecosystem-wg#17]
The [`glob`](https://crates.io/crates/glob) crate has some deficiencies that are addressed by [`globset`](https://crates.io/crates/globset), but it comes with its own [set of trade-offs](https://github.com/rust-lang-nursery/glob/issues/59#issuecomment-296852146). The [`globwalk`](https://crates.io/crates/globwalk) crate offers a `glob` function, based on `globset`, that's a missing piece of the `globset` API.

There's a willingness to [deprecate `glob` in favour of `globset`](https://github.com/rust-lang-nursery/glob/issues/59#issuecomment-297238603) and accept its trade-offs as facts of life that can be improved by the compiler.

We probably need someone to take up the cause and help push this forward.

**Comment by BurntSushi** [comment]
A possible path forward here (although, my personal bandwidth for this is super limited):

* Don't deprecate `glob`, but re-purpose it. The name is ideal. Instead, we might consider deprecating `globset` and moving its API wholesale into `glob`. Then we produce a new breaking `glob` release. This path is useful if we care about moving the ecosystem over. It's less useful if we don't care about that as much.
* My suspicion is that folks expect "globbing" and "glob the file system" to live very close together. It would be nice if they were in the same crate. I had some [thoughts here](https://github.com/BurntSushi/ripgrep/pull/765#issuecomment-361394668) previously. In particular, whatever the case, it should be possible to get "globbing" without the dependencies required for walking the file system.

I think a key issue here is that "glob the file system" needs to be proved out and polished.

At a higher level, it's also worth saying that `globset` will live on and get maintenance so long as ripgrep is alive, so I do think it makes sense to take advantage of that!

**Comment by KodrAus** [comment]
Thanks for your thoughts @BurntSushi!

Personally, I'm in favour of re-purposing the `glob` crate as `globset`'s current API, and shifting libraries into places where they get greater than zero on-board maintainers is what this is all about! :)

I'm guessing we're not in a big hurry to do this, so could work towards proving and polishing the _glob the filesystem_ case in `globwalk` before repurposing `glob`? The main point [from @BurntSushi's list](https://github.com/BurntSushi/ripgrep/pull/765#issuecomment-361394668) for me is the first one about nailing down a specification for how traversal should behave that gives us room to optimise the implementation. Maybe that's the next step here?

@Gilnaa Do you have any thoughts to share? I know it hasn't actually been very long since you started working on `globwalk`.

**Comment by KodrAus** [comment]
@Gilnaa I agree it would be a bit strange to do that. I'm assuming we'd effectively deprecate both `globset` and `globwalk`, by combining them in `glob` with a feature gate for directory traversal.

It would be good if we could get concrete about what _proving `globwalk`'s API sufficient_ means.

**Comment by Gilnaa** [comment]
Well, globset's API is a superset of glob's, so it's probably not that big of a deal; but if we could get some community feedback about it would be great.

**Comment by kinggoesgaming** [comment]
I am in favour of `glob` being repurposed and the other two crates being deprecated

**Comment by Dylan-DPC-zz** [comment]
Yeah I am in favour of having a single crate for both. We can use feature gates in case people would want only 1 of the 2 functionalities.

**Comment by BatmanAoD** [comment]
If `glob` were repurposed, and we maintained separate crates for *just globbing* (i.e. `globset`) and *globbing the filesystem* (`globwalk`), my vote would be to repurpose `glob` as a `globwalk` alternative (either based on `globset` + traversal, or simply based on the existing `globwalk`), and leave `globset` with its current name.
